### DISCUSSION TITLE
QR Code Scanner / Barcode Scanner

### DISCUSSION THREAD

**Discussion by pri-kise** [microsoft/ALAppExtensions#15184] [Ideas]
Hi @JesperSchulz ,

Recently there was a smalltalk on twitter about adding a QR Code Scanner in the system application.

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Would be awesome! Do I hear a PR coming up on GitHub? 🥳 <a href="https://twitter.com/hashtag/TogetherWeWin?src=hash&amp;ref_src=twsrc%5Etfw">#TogetherWeWin</a>! Join my session at 11:00 at Directions tomorrow, if you need a longer pep-talk than this one 😆</p>&mdash; Jesper Schulz-Wedde (MSFT) (@JesperSchulz) <a href="https://twitter.com/JesperSchulz/status/1453310938644426752?ref_src=twsrc%5Etfw">October 27, 2021</a></blockquote>

I experiemented a little bit with the following javascript library and get it to work on the webclient
https://github.com/mebjas/html5-qrcode

I thought about adding it as PR, but then I realized that the implentation via javascript isn't working in the Business Central App for Android or IOS.

I guess it's related that the app does populate the request for a camera device.
Since the library works very well in recoginizing it would be very easy to implement an AL wrapper to scan something.
Maybe you can redirect me to someone who has experience with requesting camera access through javascript in the BC Android/Ios App.

An alternative, but more complicated approach would be to create a Custom BC DLL, similar to Microsoft.Dynamics.Nav.Client.Capabilities.CameraProvider,  that will use the ZXing.Net library and then try to recoginize a BarCode 
https://github.com/micjahn/ZXing.Net

Maybe you can give some feedback, if there are any plans for this, that MS will provide something or show me which implentation could be possible?

Related Idea: https://experience.dynamics.com/ideas/idea/?ideaid=227a8464-c4e1-e811-b96f-0003ff689aeb

**Comment by JesperSchulz** [comment]
Sorry for the late reply to this discussion - I managed to build up quite the backlog of cases to reply to ;-)
Let me ask around if we've got anyone here who knows a thing or two about camera access through JavaScript - as I surely don't!

**Reply by pri-kise** [reply]
Thanks, looking forward to your next reply :)

**Reply by JesperSchulz** [reply]
I've sent a mail to a couple of developers - let's see if anyone gets back to me on this one :-) If not, can you not split the thing into two? I haven't at all looked into the details of the code you're referencing, but would it not be possible to first take the picture with the controls we've got and then do the processing of that picture afterwards, rather than doing it in one go? Just a thought.

**Reply by pri-kise** [reply]
I had this idea too, but i give up with this idea, since it can very easily happen, that the first, 2nd or 3rd picture cannot be recognized. 
Then I would have to trigger every time, taking a new picture. If the user only have to allow taking a video stream and analyzing the videos this would be much less annoying for the later user.

Here is my small demo for using the javascript library.
https://github.com/pri-kise/ALBarcodeScanner
Everything works fine within the browser on desktop computers using the notebook's camera, but sadly inside the BC App everything keeps empty (I guess missing app permissions).

**Comment by hoovercj** [comment]
Hi @pri-kise, I installed your extension and it worked fine for me on iOS 15 and Android 12 with one special caveat for Android: I had to do something else to grant camera permissions for the Android app.

For iOS, every time I triggered the barcode action, it prompted me to allow the camera and when I chose "allow", it showed a camera feed.

For Android, I could either go in to the Business Central permissions and manually allow the camera OR I could do something within the app that already triggers camera permissions, such as going to a customer card and taking a picture. After doing either of those actions, every time I triggered the barcode action I was able to see the camera feed on your barcode page.

I don't currently know of a way to force Android to prompt for camera permissions when using javascript from control-addins, but it looks like using the Html5Qrcode api or using the browser's `navigator.mediaDevices.getUserMedia` api directly, you should be able to detect scenarios where permissions are not granted and then display text guiding users to enabling the permissions in the settings and then trying again.

I hope that helps!

**Reply by pri-kise** [reply]
Hi @hoovercj, thanks for testing.
I will have to test it again on my device. I hoped there would be a way to trigger the camera permissions, similiar how it works in the browser.

**Reply by pri-kise** [reply]
I have tested it on my pixel 5 phone and after I allowed the business central app to use the camera and restarted the app, now the scanner is opened. 
Now we only would need a way to trigger the get access dialog.
@JesperSchulz do you think that this feature could be added in PR for the System Application.
I'm thinking of a Codeunit as Interface for other apps. 
And then open the page and gives a result if a Code could be read.
I only don't know how any test could be written to test the control addin.

I'm asking at first, since I don't want to invest time for something that will never be merged into the system application.
Or maybe any team is already cooking on more efficient reading with .net ....

**Reply by JesperSchulz** [reply]
I don't see why we wouldn't allow it to get added to the system application - unless there are issues with security or performance. I know that there are numerous ideas with many votes on them, meaning that the work is in our backlog somewhere. But I speak from experience when I say, that it can take many releases, before we get to that slice. If you've got a solution which either makes our work redundant, or bridges the gap while we wait for native solution, then I'd say "go for it"!
